layer1/transcation/src/transaction_sc.mligo

```diff
@@ -0,0 +1,148 @@
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE. *)
+
+#import "transaction.mligo" "Tx"
+#include "../../commons/transaction.mligo"
+
+(** This contract provides three entries: [Transaction],
+    [Transaction_from], and [Transaction_to].
+
+    When either entries is called, the transaction operation,
+    [init_state(s)] and [expected_state(s)] are needed to provide.
+    This contract will base on transaction operation and
+    [init_state(s)] to evaluate new state, compare the new state
+    with [expected_state(s)] and set compared result in storage.
+
+    [Transaction] is equal to [Transaction_from] and
+    [Transaction_to]. [Transaction_from] and [Transaction_to]
+    are for performing small steps.
+
+    Note that
+      - currently, transaction operation is only support
+        transfer tokens.
+      - this version doesn't support merkle proof.
+      - transaction operaion doesn't require to sign.
+*)
+
+
+module Result = Tx.Result
+
+(* [is_expected_state] is for storage type.
+   [Nothing] should be the default value of this contract.
+   When we perform either entres of this contract, if the
+   evaluated result is the same as [expected_state(s)], [Yes]
+   will be set. Otherwise, [No] will be set. See [storage] type*)
+type is_expected_state
+  = Yes
+  | No of { error : string; }
+  | Nothing
+
+(** storage type *)
+(** [FIXME]: support multiple game *)
+type storage =
+  {  is_expected_state : is_expected_state; }
+
+(** parameter of Transaction *)
+type tx_parameter =
+  { tx : transaction
+  ; init_states : chusai_states
+  ; expected_states : chusai_states
+  ;
+  }
+
+(** parameter of small step of transaction *)
+type 't tx_step_parameter =
+  { tx : 't
+  ; init_state : chusai_state
+  ; expected_state : chusai_state
+  ;
+  }
+
+(** parameter *)
+type parameter
+  = Transaction of tx_parameter
+  | Transaction_from of (transaction_from) tx_step_parameter
+  | Transaction_to of (transaction_to) tx_step_parameter
+
+(** return *)
+type return = operation list * storage
+
+(** compare a actual state and a expected states
+    FIXME: replace with merkletree
+*)
+let check_expected_state (result : Result.t) (expected_state : chusai_state) : is_expected_state =
+  match result with
+  | Error e -> No { error = Result.error_to_string e; }
+  | Ok state -> (
+      let b_expected_state = Bytes.pack expected_state in
+      let b_actual_state = Bytes.pack state in
+      if b_expected_state = b_actual_state then Yes
+      else No { error = "Unexpected transited state" })
```

Comment:
I think something like `"computed state different from expected state"` would be more clear. Or maybe even `"proof verification failed: computed state different from expected state"` ?